File: frontend/src/utils/cache.js
/**
 * Simple client-side cache for API responses
 */

class ClientCache {
  constructor() {
    this.cache = new Map();
  }

  /**
   * Generate cache key from URL and params
   */
  generateKey(url, params = {}) {
    const paramString = JSON.stringify(params);
    return `${url}:${paramString}`;
  }

  /**
   * Get cached value if not expired
   */
  get(key) {
    const cached = this.cache.get(key);
    if (!cached) return null;

    const { value, expiry } = cached;
    if (Date.now() > expiry) {
      this.cache.delete(key);
      return null;
    }

    return value;
  }

  /**
   * Set value in cache with TTL
   */
  set(key, value, ttl = 300000) {
    // Default TTL: 5 minutes (300000ms)
    const expiry = Date.now() + ttl;
    this.cache.set(key, { value, expiry });
  }

  /**
   * Clear all cache
   */
  clear() {
    this.cache.clear();
  }

  /**
   * Remove expired entries
   */
  cleanup() {
    const now = Date.now();
    for (const [key, { expiry }] of this.cache.entries()) {
      if (now > expiry) {
        this.cache.delete(key);
      }
    }
  }
}

// Global cache instance
export const clientCache = new ClientCache();

// Cleanup expired entries every 5 minutes
setInterval(() => {
  clientCache.cleanup();
}, 300000);

export default clientCache;

File: backend/utils/cache.py
"""
Simple in-memory cache for API responses
"""
import time
from threading import Lock

class SimpleCache:
    """Thread-safe in-memory cache with TTL support"""
    
    def __init__(self):
        self._cache = {}
        self._lock = Lock()
    
    def get(self, key):
        """
        Get value from cache if it exists and hasn't expired
        
        Args:
            key (str): Cache key
            
        Returns:
            Value if found and not expired, None otherwise
        """
        with self._lock:
            if key in self._cache:
                value, expiry = self._cache[key]
                if time.time() < expiry:
                    return value
                else:
                    # Remove expired entry
                    del self._cache[key]
            return None
    
    def set(self, key, value, ttl=300):
        """
        Set value in cache with TTL
        
        Args:
            key (str): Cache key
            value: Value to cache
            ttl (int): Time to live in seconds (default: 300 = 5 minutes)
        """
        with self._lock:
            expiry = time.time() + ttl
            self._cache[key] = (value, expiry)
    
    def clear(self):
        """Clear all cache entries"""
        with self._lock:
            self._cache.clear()
    
    def cleanup_expired(self):
        """Remove all expired entries"""
        with self._lock:
            current_time = time.time()
            expired_keys = [
                key for key, (_, expiry) in self._cache.items()
                if current_time >= expiry
            ]
            for key in expired_keys:
                del self._cache[key]

# Global cache instance
cache = SimpleCache()
